webapp/src/DOMJudgeBundle/Controller/API/AwardsController.php

```diff
@@ -0,0 +1,152 @@
+use DOMJudgeBundle\Service\DOMJudgeService;
+use DOMJudgeBundle\Service\EventLogService;
+use DOMJudgeBundle\Service\ScoreboardService;
+use FOS\RestBundle\Controller\Annotations as Rest;
+use Swagger\Annotations as SWG;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\Exception\AccessDeniedHttpException;
+
+/**
+ * @Rest\Route("/api/v4/contests/{cid}/awards", defaults={ "_format" = "json" })
+ * @Rest\Prefix("/api/contests/{cid}/awards")
+ * @Rest\NamePrefix("scoreboard_")
+ * @SWG\Tag(name="Scoreboard")
+ * @SWG\Parameter(ref="#/parameters/cid")
+ * @SWG\Response(response="404", ref="#/definitions/NotFound")
+ * @SWG\Response(response="401", ref="#/definitions/Unauthorized")
+ */
+class AwardsController extends AbstractRestController
+{
+    /**
+     * @var ScoreboardService
+     */
+    protected $scoreboardService;
+
+    /**
+     * ScoreboardController constructor.
+     * @param EntityManagerInterface $entityManager
+     * @param DOMJudgeService        $DOMJudgeService
+     * @param EventLogService        $eventLogService
+     * @param ScoreboardService      $scoreboardService
+     */
+    public function __construct(
+        EntityManagerInterface $entityManager,
+        DOMJudgeService $DOMJudgeService,
+        EventLogService $eventLogService,
+        ScoreboardService $scoreboardService
+    ) {
+        parent::__construct($entityManager, $DOMJudgeService, $eventLogService);
+        $this->scoreboardService = $scoreboardService;
+    }
+
+    /**
+     * Get the awards standings for this contest
+     * @param Request $request
+     * @return array
+     * @Rest\Get("")
+     * @SWG\Response(
+     *     response="200",
+     *     description="Returns the current teams qualifying for each award"
+     * )
+     * @throws \Exception
+     */
+    public function getAwardsAction(Request $request)
+    {
+        $public   = false;
+        if ($this->DOMJudgeService->checkrole('jury') && $request->query->has('public')) {
+            $public = $request->query->getBoolean('public');
+        }
+
+        $contest       = $this->entityManager->getRepository(Contest::class)->find($this->getContestId($request));
+        $isJury        = $this->isGranted('ROLE_JURY');
+        $accessAllowed = ($isJury && $contest->getEnabled()) || (!$isJury && $contest->isActive());
+        if (!$accessAllowed) {
+            throw new AccessDeniedHttpException();
+        }
+
+        $b = $contest->getB() ?? 0;
+
+        $scoreboard = $this->scoreboardService->getScoreboard($contest, !$public, null, true);
+
+        $group_winners = $problem_winners = [];
+        foreach ($scoreboard->getTeams() as $team) {
+            if ( $scoreboard->isBestInCategory($team) ) {
+                $group_winners[$team->getCategoryId()][] = $team->getTeamid();
+            }
+            foreach($scoreboard->getProblems() as $problem) {
+                if ($scoreboard->solvedFirst($team, $problem)) {
```

Comment:
The only problem I have with that is that this is not guaranteed to be stable. If we still are judging the submission of another team that submitted earlier but takes longer to judge we will do it wrong (see last year's WF the first correct submission).

Reply:
I think it makes sense if the awards endpoint is consistent with what you see on the scoreboard. And I think we now paint the currently first already judged correct submission datkgreen. If we want to change that, I think it needs to happen in the ScoreboardService so all output formats do it the same.

Reply:
Agreed with both.
Also there is an issue already raised (not sure by whom) that the winner may need to be temporarily undefined when a submission is being judged that could bring a new team to first place. Although we may add an exception for that in the specification.

Reply:
I think those are different things. The concept of 'winner' (probably leader is more appropriate) is expected to change continuously during the contest. I think it makes sense that we report as 'winner' at any given time whoever is the first on the scoreboard. On the scoreboard, we don't remove ranks while something is being judged (because that would essentially be always since submissions are continuously flowing in - and why would indeed rank = 3  be differnt from rank =1). I think it's sufficiently clear that all ranks are provisional until the contest has been finalized.

This is different from first-to-solve. Those awards are expected to be final as soon as they are issued, that is, if a problem has been deemed as solved first by team A, it will not later be awarded to team B (excluding rejudgings). So here it makes more sense here not to award it before it is certain that there cannot be an earlier correct submission for this problem.

Reply:
In any case this discussion is not specific to this PR, as I believe the output of the API should match what you see on the scoreboard and get from the scoreboard API.